nomt/src/new_seek.rs

```diff
@@ -0,0 +1,421 @@
+const SINGLE_PAGE_REQUEST_INDEX: usize = usize::MAX;
+
+struct SeekRequest {
+    key: KeyPath,
+    position: TriePosition,
+    page_id: Option<PageId>,
+    siblings: Vec<Node>,
+    state: RequestState,
+}
+
+impl SeekRequest {
+    fn new(key: KeyPath, root: Node) -> SeekRequest {
+        let state = if trie::is_terminator(&root) {
+            RequestState::Completed(None)
+        } else {
+            RequestState::Seeking(root)
+        };
+
+        SeekRequest {
+            key,
+            position: TriePosition::new(),
+            page_id: None,
+            siblings: Vec::new(),
+            state,
+        }
+    }
+
+    fn is_completed(&self) -> bool {
+        match self.state {
+            RequestState::Seeking(_) => false,
+            RequestState::Completed(_) => true,
+        }
+    }
+
+    fn next_page_id(&self) -> PageId {
+        match self.page_id {
+            None => ROOT_PAGE_ID,
+            // UNWRAP: all page IDs for key paths are in scope.
+            Some(ref page_id) => page_id
+                .child_page_id(self.position.child_page_index())
+                .unwrap(),
+        }
+    }
+
+    fn continue_seek(
+        &mut self,
+        read_pass: &ReadPass<ShardIndex>,
+        page_id: PageId,
+        page: &Page,
+        record_siblings: bool,
+    ) {
+        let RequestState::Seeking(mut cur_node) = self.state else {
+            panic!("seek past end")
+        };
+
+        // terminator should have yielded completion.
+        assert!(!trie::is_terminator(&cur_node));
+        assert!(self.position.depth() as usize % DEPTH == 0);
+
+        // don't set this when cur node is leaf, as we loaded this page just to get the leaf
+        // children.
+        if !trie::is_leaf(&cur_node) {
+            self.page_id = Some(page_id);
+        }
+
+        // take enough bits to get us to the end of this page.
+        let bits = self.key.view_bits::<Msb0>()[self.position.depth() as usize..]
+            .iter()
+            .by_vals()
+            .take(DEPTH);
+
+        for bit in bits {
+            if trie::is_leaf(&cur_node) {
+                let children = if self.position.depth() as usize % DEPTH > 0 {
+                    self.position.child_node_indices()
+                } else {
+                    // note: the only time this is true in this loop is in the first iteration,
+                    // because we only take `DEPTH` bits.
+                    ChildNodeIndices::next_page()
+                };
+
+                self.state = RequestState::Completed(Some(trie::LeafData {
+                    key_path: page.node(&read_pass, children.left()),
+                    value_hash: page.node(&read_pass, children.right()),
+                }));
+                return;
+            }
+
+            self.position.down(bit);
+
+            cur_node = page.node(&read_pass, self.position.node_index());
+            if record_siblings {
+                self.siblings
```

Comment:
Nit: Should we preallocate memory here with `.reserve` for the expected number of sibilings (like say 40, should be in a commented constant though)? We don't want to preallocate space when we are constructing this cuz of laziness.

Reply:
Yes, improving sibling handling w.r.t. allocations is something for an optimization pass!